PEP 771: Amendments following pre-publish discussion

astrofrog opened the pull request:
There has been an extensive discussion going on in this [DPO](https://discuss.python.org/t/pre-publish-pep-711-default-extras-for-python-software-packages/77892) thread while the original PR for PEP 771 was being reviewed. Rather than continuously update that PR, @pradyunsg suggested that that PEP should be published as-is initially and that changes based on the pre-publish discussion should be made in a separate PR. This is the separate PR :tada: 

The main changes are:

* Real-life examples of packages have been added in several places by popular demand
* Most of the 'How to teach this' section has been moved to a new 'Examples of usage' section, and 'How to teach this' has been filled with new content
* The backward-compatibility section has been expanded to address concerns raised during the discussion
* An open issue has been added as it is an unresolved point that will require more discussion
* The ``default-optional-dependencies`` key has been renamed to ``default-optional-dependency-keys``

Once this PR is merged, I'll then open the main discussion thread and open a trivial PR to add the link to the PR.

cc @pradyunsg @DEKHTIARJonathan

<!-- readthedocs-preview pep-previews start -->
----
📚 Documentation preview 📚: https://pep-previews--4238.org.readthedocs.build/

<!-- readthedocs-preview pep-previews end -->

AA-Turner commented on peps/pep-0771.rst:
```diff
@@ -41,7 +41,26 @@ control can use ``package``. However, in practice, many users are unaware
 typical installation. Having a way to have recommended dependencies be installed
 by default while providing a way for users to request a more minimal installation
-would satisfy this use case. The present PEP will describe a solution for this use case.
+would satisfy this use case, and the present PEP will describe a solution to this.
```
```suggestion
would satisfy this use case, and this PEP describes a solution to this.
```

AA-Turner commented on peps/pep-0771.rst:
```diff
@@ -41,7 +41,26 @@ control can use ``package``. However, in practice, many users are unaware
-would satisfy this use case. The present PEP will describe a solution for this use case.
+would satisfy this use case, and the present PEP will describe a solution to this.
+
+Real-world examples of packages that encourage users to include extra
+dependencies by default include for example:
```
```suggestion
Real-world examples of packages that encourage users to include extra
dependencies by default include:
```

AA-Turner commented on peps/pep-0771.rst:
```diff
@@ -41,7 +41,26 @@ control can use ``package``. However, in practice, many users are unaware
+Real-world examples of packages that encourage users to include extra
+dependencies by default include for example:
+
+* `astropy <https://www.astropy.org/>`_, a core package for astronomy, which
+  recommends that users install ``astropy[recommended]``. Some of the
+  dependencies in ``recommended``, such as `scipy <http://www.scipy.org>`_, are
+  used in a number of places in astropy for common
+  functionality and scipy should ideally be a required dependency, but there has
+  been reluctance to so as it is not a lightweight dependency and advanced users
+  who want minimal installs want the ability to opt out of it (however, as
+  described above, this places a burden on the majority of regular uses to
+  satisfy a use case for advanced users).
+* `fastapi <https://fastapi.tiangolo.com/>`_, a web framework, which recommends
+  that users install ``fastapi[standard]`` in order to include commonly needed
+  dependencies.
+* `tensorflow <https://www.tensorflow.org>`_, a machine learning platform, which
+  recommends that users install ``tensorflow[and-cuda]`` since CUDA support is
+  commonly needed by users.
```
This section feels a bit wordy, and repeats the prose description from this section. Perhaps you could list the packages, and say that these fit the description in the paragraph above?

E.g. 

```rst
There are several packages that demonstrate this pattern by encouraging users to include extra dependencies by default, such as:

* (link to astropy): ``astropy[...]``
```

astrofrog replied:
Done

AA-Turner commented on peps/pep-0771.rst:
```diff
@@ -70,6 +89,32 @@ or backend package. There is currently no mechanism in Python packaging
 infrastructure to disallow conflicting or incompatible extras to be installed,
 and this PEP does not change that.
 
+Real-world examples of packages that allow users to select different backends or
+frontends with extras include:
```
Likewise, I wonder if there's a better (more concise) way of integrating the description of the pattern with the examples.

astrofrog replied:
Done

AA-Turner commented on peps/pep-0771.rst:
```diff
@@ -79,9 +124,9 @@ the community for several years, including in `this DPO thread
 as well as in numerous issues and pull requests. The solution that will be
 presented below:
```
```suggestion
as well as in numerous issues and pull requests. The solution that is
presented below:
```

AA-Turner commented on peps/pep-0771.rst:
```diff
@@ -79,9 +124,9 @@ the community for several years, including in `this DPO thread
 * is an opt-in solution which means that package maintainers can choose whether or not to use it
-* is flexible enough to accommodate both of the major use cases described in `Motivation`_.
+* is flexible enough to accommodate both of the major use cases described in `Motivation`_
+* does not introduce new syntax compared to :pep:`508` (which would have serious backward-compatibility implications)
```
"serious backward-compatibility implications" here is unqualified, I would expect an explanation or a link to a section where it is justified.
```suggestion
* re-uses the syntax from :pep:`508`
```

AA-Turner commented on peps/pep-0771.rst:
```diff
@@ -202,34 +260,84 @@ extras.
-All package specification cases valid under :pep:`508` will remain valid.
-Therefore, this proposal is fully backward-compatible with existing :pep:`508`
-usage.
+There are several different aspects related to backward compatibility that need
+to be considered.
```
This sentence doesn't add anything
```suggestion
```

AA-Turner commented on peps/pep-0771.rst:
```diff
@@ -202,34 +260,84 @@ extras.
 There are no known security implications for this PEP.
 
-How to teach this
+Examples of usage
 =================
```
Please stick to the suggested sections in [PEP 12](https://peps.python.org/pep-0012/#suggested-sections). Perhaps you could add "examples of usage" as a sub-section, but "how to teach this" is broader than simply stating examples.
```suggestion
How to teach this
=================
```

AA-Turner reviewed (commented):
Brief review, I have only read up to the changes to Specification in depth.

Edit: Please can you add the linked thread to Post-History?

```rst
Post-History:
  `15-Jan-2025 <https://discuss.python.org/t/77892>`__,
```

astrofrog commented:
@AA-Turner - thanks for the review so far! I've added the ``Post-History`` as requested (and have implemented the other comments). I also moved the examples section into a sub-section of the specification (as opposed to how to teach this) since it felt more natural as the last subsection of the specification, and there is already an extensive 'how to teach this' - however if you feel strongly this should be a subsection of 'how to teach this' I can move it, so just let me know.

willingc commented on peps/pep-0771.rst:
```diff
@@ -350,11 +440,258 @@ This could be achieved with e.g:
+^^^^^^^^^^^^^^^^^^^
+
+In some cases, it may be that packages have extras that are mutually
+incompatible. In this case, we recommend against making using the default extra
+feature for any extra that contains a dependency that could be incompatible with
+another.
```
```suggestion
incompatible. In this case, we recommend against using the default extra
feature for any extra that contains a dependency that could be incompatible with
another.
```

willingc commented on peps/pep-0771.rst:
```diff
@@ -350,11 +440,258 @@ This could be achieved with e.g:
+
+.. code-block:: console
+
+    $ pip install package
```
Perhaps add a comment  after the command to specify what is being installed (A as a default)

astrofrog replied:
Done

willingc commented on peps/pep-0771.rst:
```diff
@@ -350,11 +440,258 @@ This could be achieved with e.g:
+Package users
+-------------
+
+Package users should be provided with clear installation instructions that show
```
As a napari developer that deals with qt versions and dependencies, clarity is key as a user.

astrofrog replied:
I agree - are there any specific wording improvements you would like to see here, or is this ok?

willingc replied:
No change needed. This was more a bravo for adding this!

willingc commented on peps/pep-0771.rst:
```diff
@@ -350,11 +440,258 @@ This could be achieved with e.g:
+  will need to carefully consider how to treat default extras, and this may
+  imply a non-negligible amount of work and discussion.
+
+It is impossible for a PEP such as this to exhaustively consider each of the
```
Hmm... @astrofrog, do you expect the impact to be felt most by package maintainers, packaging tool providers, packaging repositories, or package users? It's not clear to me.

astrofrog replied:
It's a bit of a tricky question - if done properly, there shouldn't be a negative impact for end users.

For packaging tool providers, there will be the burden of implementing the PEP and then fixing bugs for corner cases over time. Hopefully once the PEP is implemented, the burden should be low-ish.

For package authors, the burden is opt-in - if they choose to use this feature they should think about it to plan carefully (especially for established packages).

For packaging repositories, I think the main impact will be for repositories that use automation to do the repackaging because there will need to be careful planning related to how to handle this PEP. I think the additional burden will be low for repositories where the repackaging is done by hand.

Sorry this isn't a simple answer :sweat_smile: - but let me know if you think I should add something along these lines to the PEP. It's not really clear where this would go though as there isn't really an 'impact' section where this could be discussed, and it's not *really* part of 'How to teach this'?

willingc replied:
I think that maybe the way to handle this is to add an intro sentence to "How to teach" such as:

```
This section discusses the information and preparations someone may need to consider after implementing this PEP. The groups covered are:
- End users (package consumers)
- Packaging tooling providers
- Package authors
- Packaging repository maintainers
```
And link to the headings that you have (or could have) for each of these groups. Thoughts?

astrofrog replied:
@willingc - I've now added an intro paragraph to this section based on your suggestion above

willingc reviewed (commented):
Thanks for taking this on @astrofrog. I have several comments and questions after reading the PEP but not all of the DPO discussion. I agree with the Motivations, but I'm wrestling a bit if we're communicating the approach clearly enough. I'm not advocating changing the approach but making it more explicit for readers.

pradyunsg commented on peps/pep-0771.rst:
```diff
@@ -156,34 +181,47 @@ extras are not installed. Otherwise, the default extras are used.
 Ideas`_ section for the rationale).
 
+We also note that some tools such as `pip`_ currently ignore unrecognized
+extras, and emit a warning to the user to indicate that the extra has not been
```
There are no concept of default extras today, so the non-default qualifier doesn't add any context to this sentence that is about status quo.

Maybe an example of `pip install package[this-extra-does-not-exist-in-metadata]` with the warning printed by pip would make this clearer.

astrofrog replied:
I've added an example of pip emitting a warning

pradyunsg commented on peps/pep-0771.rst:
```diff
@@ -350,11 +440,258 @@ This could be achieved with e.g:
+  ``minimal`` in 2.0, then downstream users and packages that want to depend on
+  a minimal version of the package cannot declare the following dependency::
+
+    packaage[minimal]>=1.0
```
```suggestion
    package[minimal]>=1.0
```

astrofrog replied:
Thanks! @willingc made the same suggestion and I committed hers

astrofrog commented:
@willingc - thank you for your review! I will try and implement the comments by early next week.

astrofrog commented on peps/pep-0771.rst:
```diff
@@ -156,34 +181,47 @@ extras are not installed. Otherwise, the default extras are used.
-    pip install package[extra2]
+.. code-block:: console
+
+    $ pip install package[extra2]
```
Done

astrofrog commented on peps/pep-0771.rst:
```diff
@@ -156,34 +181,47 @@ extras are not installed. Otherwise, the default extras are used.
+.. code-block:: console
 
-    pip install package package[extra2]
+    $ pip install package package[extra2]
```
Done - I agree this command isn't easy to parse visually, I've just replaced this with a dependency tree now.

astrofrog commented on peps/pep-0771.rst:
```diff
@@ -156,34 +181,47 @@ extras are not installed. Otherwise, the default extras are used.
 
+We also note that some tools such as `pip`_ currently ignore unrecognized
+extras, and emit a warning to the user to indicate that the extra has not been
+recognized. In this case, if no extras are recognized as being valid in a
```
Reworded along these lines

astrofrog commented on peps/pep-0771.rst:
```diff
@@ -199,40 +237,18 @@ be installed but will not work if none are provided, so a package maintainer may
-The rule above regarding only installing default extras when no extras
-are explicitly specified, combined with the introduction of the
-``Default-Extra:`` keyword and ``default-optional-dependencies`` metadata key
+The rule in `Overriding default extras`_ regarding only installing default extras
```
I've shortened this paragraph significantly

astrofrog commented on peps/pep-0771.rst:
```diff
@@ -277,8 +293,41 @@ Maintainers would have the choice as to whether to offer the capability to do a
 minimal installation or not - in some cases, such as highlighted in the next
 section, this might not be desirable.
 
+To take one of the concrete examples from the `Motivation`_ section, the
```
Done - the examples in this and the next two sub-sections are now self-contained (while still referring to the Motivation to indicate they were mentioned before)

astrofrog commented on peps/pep-0771.rst:
```diff
@@ -350,11 +440,258 @@ This could be achieved with e.g:
+
+Once support for this PEP is added to tools in the packaging ecosystem, packages
+that do not make use of default extras will continue to work as-is and there
+should be no break in compatibility.
```
Yes that is correct. Would you like to see the wording changed here?

astrofrog commented on peps/pep-0771.rst:
```diff
@@ -350,11 +440,258 @@ This could be achieved with e.g:
+with recent versions of packaging tools which implement this PEP, but those
+packages will remain installable with older packaging tools -- with the main
+difference being that the default extras will not be installed automatically
+when older packaging tools are used. As described in `How to teach this`_,
```
Done

astrofrog commented:
@willingc -  thanks again for the review! I've tried to implement your comments - there are a couple of remaining questions in some of the in-line discussions regarding whether any further text needs to be added or modified, so just let me know what you think.

willingc commented:
Skimming through this @astrofrog, I think that you have captured well the edits. :smile: I will take a look at the inline later tomorrow. Thank you!

astrofrog commented:
This branch is now out of date with main - should I just add a merge commit to fix that? (in the last PR I rebased but was told it wasn't ideal for reviews so just want to make sure I do the right thing)

hugovk commented:
It's fine, there's no need to update from `main`, unless there's something you specifically need from there (or maybe if this PR had been open a long time and we want to check if the CI still passes).

willingc reviewed (approved):
Beautiful. Thanks @astrofrog.

astrofrog commented:
Thanks for the approval! Once this is merged, I will open the main discussion thread on DPO and will open a follow-up PR updating the discussion link.

hugovk commented on peps/pep-0771.rst:
```diff
@@ -41,7 +43,14 @@ control can use ``package``. However, in practice, many users are unaware
+Examples of packages that demonstrate this pattern by encouraging users to
+include extra dependencies by default include:
+
+* `astropy <https://www.astropy.org/>`_: ``astropy[recommended]``.
```
nit
```suggestion
* `astropy <https://www.astropy.org/>`_: ``astropy[recommended]``
```

hugovk commented on peps/pep-0771.rst:
```diff
@@ -70,18 +82,31 @@ or backend package. There is currently no mechanism in Python packaging
+
+* `napari <https://napari.org>`_: ``napari[all]``
+
+* `glueviz <https://www.glueviz.org>`_: ``glueviz[qt]``.
```
```suggestion
* `glueviz <https://www.glueviz.org>`_: ``glueviz[qt]``
```

hugovk reviewed (commented):
Any final comments from @pradyunsg? Otherwise let's merge.

pradyunsg commented:
> Any final comments from @pradyunsg?

Nope. :)

AA-Turner commented on peps/pep-0771.rst:
```diff
@@ -6,6 +6,8 @@ Status: Draft
 Topic: Packaging
 Created: 13-Jan-2025
+Post-History:
+  `15-Jan-2025 <https://discuss.python.org/t/77892/>`__,
```
Do we want to include the topic to be created now here, just before merging?

astrofrog replied:
If rather do it the other way around if possible, that is do a follow up small PR with the thread otherwise if I make the new thread now it will temporarily link to the old version of the PEP and could cause confusion?

astrofrog replied:
Or is there a way to reserve a thread URL before actually making it public?

AA-Turner replied:
> temporarily link to the old version of the PEP

For no more than a minute, though.

> reserve a thread URL before actually making it public?

You can create a thread with only a title but no text, and edit the text in later. But this is probably overkill. If you're really worried, you can create the thread with the [preview link](https://pep-previews--4238.org.readthedocs.build/) and edit it to https://peps.python.org/pep-0771/ the moment this is merged?

A

astrofrog replied:
> For no more than a minute, though.

Sorry maybe there's a subtlety that I don't understand, but this will rely on you or someone else merging the PR as soon as I add the commit here to add the link to the thread, so in principle could be a much longer time? Or are you saying you are ready to merge at short notice? 😆

astrofrog commented:
@AA-Turner - I've made the thread and have added Discussion-To here - would you or someone else be able to merge shortly? Thanks!

AA-Turner commented on peps/pep-0771.rst:
```diff
@@ -2,10 +2,13 @@ PEP: 771
 Title: Default Extras for Python Software Packages
 Author: Thomas Robitaille <thomas.robitaille@gmail.com>, Jonathan Dekhtiar <jonathan@dekhtiar.com>
 Sponsor: Pradyun Gedam <pradyunsg@gmail.com>
+Discussions-To: https://discuss.python.org/t/pep-771-default-extras-for-python-software-packages/79706
```
```suggestion
Discussions-To: https://discuss.python.org/t/79706/
```

AA-Turner commented on peps/pep-0771.rst:
```diff
@@ -2,10 +2,13 @@ PEP: 771
 Topic: Packaging
 Created: 13-Jan-2025
+Post-History:
+  `15-Jan-2025 <https://discuss.python.org/t/77892/>`__,
```
```suggestion
  `15-Jan-2025 <https://discuss.python.org/t/77892/>`__,
  `06-Feb-2025 <https://discuss.python.org/t/79706/>`__,
```

eli-schwartz commented on peps/pep-0771.rst:
```diff
@@ -350,11 +460,304 @@ This could be achieved with e.g:
+authors of ``package2``.
+
+
+Packaging repository maintainers
+--------------------------------
+
+The impact on individuals who repackage Python libraries for different
+distributions, such as `conda <https://docs.conda.io>`_, `Homebrew
+<https://brew.sh/>`_, linux package installers (such as ``apt`` and ``yum``) and
+so on, needs to be considered. Not all package distributions have mechanisms
+that would line up with the approach described. In fact, some distributions such
+as conda, do not even have the concept of extras.
```
Speaking from "the other side", I would have to say that pip functionality is a strict subset of the functionalities available by repackagers. Primarily based on my knowledge of linux package managers, but also with regard to conda and spack. I know you're specifically calling out conda here as not being able to support the PEP, but I'm saying conda *specifically* can -- and so can everyone else.

I feel like the python community by and large has a lot of expertise in how pip works but very little study into how anything other than pip works. :( There is a lot of interesting practical research that's been done over the decades. Perhaps you're already planning to do that though -- the wording of this section somehow gives me the feeling that it's a bit of an open TODO to flesh out the actual considerations! Feel free to ask me for more details if so.

(Incidentally: I oppose the PEP anyway because despite all the functionality of it being representable by repackagers, the PEP largely makes decisions that explicitly close the door on what I would consider as the primary functionality of "extras" as they are used in other ecosystems. There is very little to gain here without taking a holistic view of things. And much to lose if you e.g. define what is currently unspecified/invalid as newly-valid without adding new functionality using it. :P)
